### ISSUE TITLE
Changed TypeIs behavior in 1.1.361 to 1.1.364

### ISSUE THREAD

**Issue by cstruct** [microsoft/pyright#8009]
The behavior of `TypeIs` changed in version `1.1.361` and forward, i don't know if this is a regression or if non-standard behavior was previously allowed.

**Describe the bug**
`TypeIs` previously did type narrowing and currently does nothing.

**Code or Screenshots**

```python
from __future__ import annotations

from collections.abc import Iterable, Sequence
from typing import TypeVar

from typing_extensions import TypeIs, TypeVarTuple, Unpack, reveal_type

Ts = TypeVarTuple("Ts")
T = TypeVar("T")
U = TypeVar("U")


class Join(Sequence[tuple[T, Unpack[Ts]]]):
    _left: Sequence[tuple[T, Unpack[Ts]]]

    @staticmethod
    def _is_unary_instance_left(join: Join) -> TypeIs[Join[T]]:
        if (first := next(iter(join._left), None)) and len(first) == 1:
            return True
        return False

    def inner(self, right: Iterable[U]) -> Join[T, Unpack[Ts], U]:
        if self._is_unary_instance_left(self):
            # Pyright 1.1.360: Type of "self._left" is "Sequence[tuple[T@Join]]"
            # Pyright 1.1.361-1.1.364: Type of "self._left" is "Sequence[tuple[T@Join, *Ts@Join]]"
            reveal_type(self._left)
        else:
            # Pyright 1.1.360-1.1.364: Type of "self._left" is "Sequence[tuple[T@Join, *Ts@Join]]"
            reveal_type(self._left)
        raise NotImplementedError()
```

**VS Code extension or command-line**
Stated versions in the terminal, `1.1.360`, `1.1.361`, `1.1.364`.

**Comment by erictraut** [comment]
I fixed several bugs in `TypeIs` handling in 1.1.362. That explains the difference in behavior that you're seeing.

The new behavior is what I would expect in this case. It also matches the behavior that I'm seeing in mypy.

One thing that looks suspect in your code is that the return type of the type guard function is `Join[T]`, but `Join` takes two type arguments. Since you've omitted the second type argument, a type checker must assume the default value for a TypeVarTuple which is `*tuple[Any, ...]`. That's probably not what you intended. I think what you really intend here was `Join[T, *tuple[()]]`.

[PEP 742](https://peps.python.org/pep-0742/#type-narrowing-behavior), which introduced `TypeIs`, specifies the "correct" narrowing behavior based on type theory, but it also indicates that the type theoretic behavior cannot be achieved in practice given limitations in the Python type system. 

If we apply my bug fix to your code, the "correct" narrowed type in the positive ("if") case is:
`Self[T, *Ts] ∧ Join[T, *tuple[()]]` (where "∧" is a set intersection operator). There is currently no intersection support in the Python type system, so type checkers like pyright and mypy have various heuristics that approximate intersections. This particular use case is not covered by the heuristics in either pyright or mypy, so both of them fall back on the approximation of `Self[T, *Ts]` which is still correct (in spec) but perhaps not as narrow as you would like.

I'll note that mypy appears to have a bug in its negative ("else") type narrowing case here. It is narrowing the type to `Never` in the negative case, which is incorrect according to type theory and not conformant with the spec.

I'm going to close the issue because I think pyright is doing the right thing here.

**Comment by cstruct** [comment]
Thank you for the quick and thorough response. Good to hear that it wasn't a bug and just a lack of my understanding. In this case changing to `TypeGuard` and fixing the missing empty tuple unpacks solved my issue. I thought I grasped how `TypeGuard` and `TypeIs` worked but evidently I have to do some more reading and playing around with them. Thank you again!
